clang/include/clang/Basic/LangOptions.h

```diff
@@ -816,6 +816,11 @@ class FPOptions {
     setAllowFPReassociate(LO.AllowFPReassoc);
     setNoHonorNaNs(LO.NoHonorNaNs);
     setNoHonorInfs(LO.NoHonorInfs);
+    // Ensure that if FiniteMathOnly is enabled, NoHonorNaNs and NoHonorInfs are
+    // also enabled. This is because FiniteMathOnly mode assumes no NaNs or Infs
+    // are present in computations.
+    if (!LO.NoHonorInfs || !LO.NoHonorInfs)
+      assert(!LO.FiniteMathOnly && "FiniteMathOnly implies NoHonorInfs");
```

Comment:
How about:
```
assert((LO.FiniteMathOnly == (LO.NoHonorInfs && LO.NoHonorNans)) && "inf/nan inconsistent internal state");
```
to make it more clear that the state has to be sane no matter which options are set? Also, this corrects the duplicate condition in your `if` statement.

Reply:
With this condition, this command line will assert:
`clang.exe -Xclang -menable-no-infs -Xclang  -menable-no-nans foo.c`
` FiniteMathonly` is `0` by default.
Do we want that?

Reply:
That sounds like a logic bug to me -- if infs and nans are disabled, then it's definitely finite math only mode. If infs *OR* nans are disabled... I think it's also finite math only mode.

Ideally, I think we should get rid of the internal notion of `FiniteMathOnly` and replace it with the honor infs/nans flags explicitly so we don't get into these sort of situations to begin with. We'd still have to answer how to define `__FINITE_MATH_ONLY__`, but that seems doable.

CC @andykaylor @jcranmer-intel

Reply:
There is quite a bit of LIT tests that have that "logic bug" then. They are using` -Xclang -menable-no-infs -Xclang  -menable-no-nan` without the `-ffinite-math-only` option.

Getting rid of `FiniteMathOnly` is fine by me. 
`FiniteMathOnly = !HonorINFs && !HonorNaNs` which would mean `-menable-no-infs -menable-non-nans`. I would think using one or the other of these options could still mean that input values may be either INF or NAN or we can produce them, in which case we wouldn't be in finite math mode.

Reply:
> There is quite a bit of LIT tests that have that "logic bug" then. They are using -Xclang -menable-no-infs -Xclang  -menable-no-nan without the -ffinite-math-only option.

I didn't mean the logic bug was on the user end; I meant that our internal state has the logic bug because `!FiniteMathOnly` implies that infinities are supported and so it logically conflicts with `!HonorInfs`.

> Getting rid of FiniteMathOnly is fine by me.
FiniteMathOnly = !HonorINFs && !HonorNaNs which would mean -menable-no-infs -menable-non-nans. I would think using one or the other of these options could still mean that input values may be either INF or NAN or we can produce them, in which case we wouldn't be in finite math mode.

I can see the logic there and I think that's defensible, but because `FiniteMathOnly` determines the value `__FINITE_MATH_ONLY__` expands to, we should probably see how that flag is being used in the wild to see whether folks expect it to be `1` or `0` when infinity is enabled but NAN is disabled (for example).

Reply:
I agree with Aaron that we should get rid of the internal state that combines these two modes. I expect we'll only want to define `__FINITE_MATH_ONLY__` when we aren't honoring either. You could create code that made incorrect assumptions either way, but the state described by the symbol is that we aren't honoring either.

Reply:
The way I see it, nnan and ninf collectively define four modes for math: full types, no-nan, finite-math (nnan + ninf), and the extremely questionable (to me at least) ninf-without-nnan. I don't think `FiniteMathOnly` as a concept independent of the nnan/ninf flags makes much sense.

There is perhaps room for disagreement as to whether `__FINITE_MATH_ONLY__` should be `nnan || ninf` or `nnan && ninf`; I lean towards && myself. Looking at sourcegraph, there seems to be essentially three uses of these macros: two to indicate that they don't care about min(NaN) results, and one to avoid using fpclassify.